Add output parser

Aflynn50 opened the pull request:
Add the functions for parsing outputs and updated the tests.
Some of the functions are somewhat similar to each other and could possibly be merged into one however we have elected to keep them separate with the view that with future updates their definitions will diverge.

TheSignPainter98 commented on internal/parse/parser.go:
```diff
@@ -123,6 +123,13 @@ func (p *Parser) Parse(input string) (expr *ParsedExpr, err error) {
 	for {
 		p.partStart = p.pos
 
+		if op, ok, err := p.parseOutputExpression(); err != nil {
```
Naming the return value `op` here says 'operation' to me, and given that the function is called 'parseOutputExpression,' perhaps naming this `out` or `outPart` would be more intuitive.

I realise `op` is consistent with `ip`, but I think both names could benefit from change.

Aflynn50 replied:
That is true, ip is almost as confusing as op

TheSignPainter98 commented on internal/parse/parser.go:
```diff
@@ -278,6 +304,138 @@ func (p *Parser) parseGoFullName() (FullName, bool, error) {
+			if p.skipByte(')') {
+				return objs, true, nil
+			}
+			nextItem = p.skipByte(',')
```
Should you skip spaces before the `','`? Having a space there would be bad formatting, but might not be considered a syntax error

Aflynn50 replied:
We do, on line 327

TheSignPainter98 replied:
327 checks for a space before the bracket (` ),`), this code doesn't allow for whitespace between the bracket and comma (`) ,`) :P

Aflynn50 replied:
I'm not sure I understand, once we parse a closing bracket we return. When the bracket is closed we finish parsing, we want to ignore everything after the bracket in this function and leave it up to the caller how they want to handle it.

TheSignPainter98 replied:
Ah right my bad I misread, this is fine

TheSignPainter98 commented on internal/parse/parser.go:
```diff
@@ -278,6 +304,138 @@ func (p *Parser) parseGoFullName() (FullName, bool, error) {
+
+	// An '&' must be preceded by a space and succeeded by a name or opening
+	// bracket.
+	if p.skipString(" &") {
```
Same note here about the number of spaces before the `&`, here this would be consistent with `AS`

Aflynn50 replied:
Ah yea, so we have in the spec that & needs to be preceded by a space. If there is more than one space that isnt a problem since we advance the parse one step at a time (i.e. the char before that space can also be a space so there can be as many as the user wants) and when PR #16 is added it will respect this as well.

TheSignPainter98 commented on internal/parse/parser.go:
```diff
@@ -278,6 +304,138 @@ func (p *Parser) parseGoFullName() (FullName, bool, error) {
+	cp := p.save()
+	// We skip a space here to keep consistent with parseTargets which also
+	// consumes one space before the start of the expression.
+	p.skipByte(' ')
```
Should it always be preceded by a single space or by a positive number of them? (As in `skipSpaces` returns `true`)

Aflynn50 replied:
This is just to keep consistant with parseTargets, since it is at the start of the expression it can have any number of spaces before it, they are just not captured as part of the expression.

solmyrcan commented on internal/parse/parser.go:
```diff
@@ -257,6 +264,25 @@ func (p *Parser) parseIdentifier() (string, bool) {
+// parseList.
+func (p *Parser) parseColumn() (FullName, bool, error) {
+	cp := p.save()
+	if id, ok := p.parseIdentifierAsterisk(); ok {
```
I might be wrong, so please check: `parseIdentifierAsterisk()` seems to return `true` if the position is at an asterisk (`*`) character.

_**Case 1:**_ `FullName: *somename`

`id` becomes `*`. The inner if block execution falls over to the `else`, and we get `Fullname: someone`

_**Case 2:**_ `FullName: *.somename`

`skipByte()` in the inner `if` returns `true`, and then we execute `parseIdentifierAsterisk()` again, which would in turn call `parseIdentifier()`.

We get: `Prefix: *, Name:somename` for the `FullName` output. 

**_My question is:_** 
At the inner `if`, can we instead call `parseIdentifier()` directly,rather than `parseIdentifierAsterisk()`? 

Or do we call `parseIdentifierAsterisk()`, because we may get a possible inner `*` as well (something like `FullName: *.*somename`) ?

However, if it is indeed possible to have an an inner `*`, then the second `parseIdentifierAsterisk()` would return `*` as well, skipping the remaining `somename`, and we get `FullName = Prefix: *, Name: *`

This suggests that we probably don't expect an inner `*` - is my interpretation correct? :)

Aflynn50 replied:
So, what we are trying to capture with this function is that a column can be: `*`, `col`, `t.col` or `t.*`. If we don't encounter a `.` then we want to put the name in the `Name` field. If we counter a `.` we want to put the first thing in the `Prefix` field and the second in the `Name` field.

There is a problem in that it is possible to have `*.*` which wouldn't be valid. But I as far as I know in SQL (I may very well be wrong) you don't get columns of the form `*col` so they shouldn't be an issue.

TheSignPainter98 commented on internal/parse/parser.go:
```diff
@@ -278,6 +304,138 @@ func (p *Parser) parseGoFullName() (FullName, bool, error) {
+
+// parseOutputExpression parses all output expressions. The ampersand must be
+// preceded by a space and followed by a name byte.
+func (p *Parser) parseOutputExpression() (op *OutputPart, ok bool, err error) {
```
If you choose to change `op` to something else, it should be updated here also :)

Aflynn50 replied:
Done.

solmyrcan commented on internal/parse/parser.go:
```diff
@@ -278,6 +304,138 @@ func (p *Parser) parseGoFullName() (FullName, bool, error) {
+			} else {
+				return []FullName{}, false, fmt.Errorf("invalid identifier near char %d", p.pos)
+			}
+			p.skipSpaces()
```
Here, `p.skipSpaces()` executes for a second time I think, is this intentional?

Aflynn50 replied:
It is, thats in case the user does `col1 , col2` which is weird but valid.

solmyrcan replied:
Actually what I meant was that after skipSpaces at line 321 executes (the case for err=nil and ok is OK :P ), execution falls to line 327, which again skips spaces. The error cases there do return(), so they cannot run line 327 either, so it feels like line 327 is redundant

Aflynn50 replied:
Ah yea, good point, so it is.

fernape commented on internal/parse/parser.go:
```diff
@@ -277,6 +303,137 @@ func (p *Parser) parseGoFullName() (FullName, bool, error) {
 
+// parseList takes a parsing function that returns a FullName and parses a
+// bracketed, comma seperated, list. On success it returns an array of FullName
+// objects. Otherwise it returns an empty array, false and an error.
```
_Otherwise it returns  a nil array..._
A `nil` array is always empty but an empty array is not necessarily `nil`. Since we explicitly return `nil`, I think the comment should reflect that.

Aflynn50 replied:
Changed.

fernape commented on internal/parse/parser.go:
```diff
@@ -277,6 +303,137 @@ func (p *Parser) parseGoFullName() (FullName, bool, error) {
+		return []FullName{col}, true
+	} else if cols, ok, _ := p.parseList((*Parser).parseColumn); ok {
+		return cols, true
+	}
```
This would be equivalent and a bit more readable maybe? We could easily add a comment for the list case.
```       
        // Case 1: A single column e.g. p.name
        if col, ok, _ := p.parseColumn(); ok {
                return []FullName{col}, true
        }

        // Case 2: Multiple columns
        if cols, ok, _ := p.parseList((*Parser).parseColumn); ok {
                return cols, true
        }
```

Aflynn50 replied:
Changed, and make the examples in the Case comments consistent.

fernape commented on internal/parse/parser.go:
```diff
@@ -277,6 +303,137 @@ func (p *Parser) parseGoFullName() (FullName, bool, error) {
+		} else if err != nil {
+			return nil, false, err
+			// Case 2: Multiple targets e.g. &(Person.name, Person.id)
+		} else if targets, ok, err := p.parseList((*Parser).parseGoFullName); ok {
```
Similarly as [above](https://github.com/canonical/sqlair/pull/23/files#r1044390489) we could remove the last `} else if...` of the first branch and make them two separate branches. The logic would be the same. In fact, that seems the pattern we have in `parseOutputExpression()`

Aflynn50 replied:
Changed, and also the one in parseOutputExpression

fernape commented on internal/parse/parser.go:
```diff
@@ -277,6 +303,137 @@ func (p *Parser) parseGoFullName() (FullName, bool, error) {
+
+// parseOutputExpression parses all output expressions. The ampersand must be
+// preceded by a space and followed by a name byte.
+func (p *Parser) parseOutputExpression() (out *OutputPart, ok bool, err error) {
```
We are not using output named parameters in other places. I'm not sure if they add much value here. For instance, line 405:
```
if targets, ok, err = p.parseTargets(); ok {
```
It seems the intention is to use `ok` and `err` as local variables and not overwrite the output parameters, since all the `return` statements use specific values.

Maybe we could remove the output names.

```
diff --git a/internal/parse/parser.go b/internal/parse/parser.go
index 6ad084e..d9c14d6 100644
--- a/internal/parse/parser.go
+++ b/internal/parse/parser.go
@@ -396,24 +396,23 @@ func starCount(fns []FullName) int {
 
 // parseOutputExpression parses all output expressions. The ampersand must be
 // preceded by a space and followed by a name byte.
-func (p *Parser) parseOutputExpression() (out *OutputPart, ok bool, err error) {
+func (p *Parser) parseOutputExpression() (*OutputPart, bool, error) {
        cp := p.save()
        var cols []FullName
-       var targets []FullName
 
        // Case 1: simple case with no columns e.g. &Person.*
-       if targets, ok, err = p.parseTargets(); ok {
+       if targets, ok, err := p.parseTargets(); ok {
                return &OutputPart{cols, targets}, true, nil
        } else if err != nil {
                return nil, false, err
        }
-       if cols, ok = p.parseColumns(); ok {
+       if cols, ok := p.parseColumns(); ok {
                // Case 2: The expression contains an AS
                // e.g. "p.col1 AS &Person.*".
                numCols := len(cols)
                p.skipSpaces()
                if p.skipString("AS") {
-                       if targets, ok, err = p.parseTargets(); ok {
+                       if targets, ok, err := p.parseTargets(); ok {
                                numTargets := len(targets)
                                // If the target is not * then check there are equal columns
                                // and targets.
 ```

Aflynn50 replied:
Changed.

niemeyer commented on internal/parse/parser.go:
```diff
@@ -273,11 +302,152 @@ func (p *Parser) parseGoFullName() (FullName, bool, error) {
+func (p *Parser) parseList(parseFn func(p *Parser) (FullName, bool, error)) ([]FullName, bool, error) {
+	cp := p.save()
+
+	if p.skipByte('(') {
```
Trivial: this might be silghtly simplified / optimized if reorganized as:
```
if !p.skipByte('(') {
    return ...
}

cp := p.save()

Aflynn50 replied:
Changed.

niemeyer commented on internal/parse/parser.go:
```diff
@@ -273,11 +302,152 @@ func (p *Parser) parseGoFullName() (FullName, bool, error) {
+			} else if err != nil {
+				return nil, false, err
+			} else {
+				return nil, false, fmt.Errorf("invalid identifier near char %d", p.pos)
```
Can't quite tell if this will work at this point, so just FYI for now: this may create issues if the expression being parsed is a normal SQL expression that is not supposed to be interpreted by SQLAir. For this not to be an error, this point must only be reached once we're sure this is a SQLAir expression.

niemeyer replied:
Also, maybe instead of "near char %d", we can use it as a prefix such as "column %d: invalid identifier"

Aflynn50 replied:
The way it currently works is that the caller of parseList decides weather to heed the error or not. If they are trying to parse something they know is an expression they can pass the error on, otherwise they will ignore it.

Changed the comment.

Aflynn50 replied:
With the new logic for parsing `(&P.a, &P.b)` this did need changing. It now works as follows:
- If the first item in the list doesn't parsed we return `ok=false, err=nil`
- If the first item did parse and any subsequent items dont/the closing bracket is missing we throw an error

niemeyer replied:
Nice, that sounds like helpful behavior.

niemeyer commented on internal/parse/parser.go:
```diff
@@ -273,11 +302,152 @@ func (p *Parser) parseGoFullName() (FullName, bool, error) {
+			nextItem = p.skipByte(',')
+			p.skipSpaces()
+		}
+		return nil, false, fmt.Errorf("missing closing parentheses for char %d", parenPos)
```
Same as above about the "column %d: " prefix.

Aflynn50 replied:
Changed.

niemeyer commented on internal/parse/parser.go:
```diff
@@ -273,11 +302,152 @@ func (p *Parser) parseGoFullName() (FullName, bool, error) {
+func (p *Parser) parseColumns() ([]FullName, bool) {
+	cp := p.save()
+
+	// We skip a space here to keep consistent with parseTargets which also
```
That seems a bit awkward. I'd review both to not skip spaces, but rather expect the preceding logic to make sure that this is the boundary of an expression and not the middle of a word. Otherwise, we'll find awkward bugs such as not supporting parenthesis when they are innocuous for instance.

Aflynn50 replied:
I have removed this, though we will only have the desired behaviour (only recognising & when it is preceded by a space) once #16 is merged.

niemeyer replied:
Ack.

niemeyer commented on internal/parse/parser.go:
```diff
@@ -273,11 +302,152 @@ func (p *Parser) parseGoFullName() (FullName, bool, error) {
+func (p *Parser) parseTargets() ([]FullName, bool, error) {
+	cp := p.save()
+
+	// An '&' must be preceded by a space and succeeded by a name or opening
```
Same note as above about the space: ideally we can organize it so that the call site is responsible for identifying whether this is a proper place to have targets or not, only call this function at the right places.

Aflynn50 replied:
Same reply as above.

niemeyer commented on internal/parse/parser.go:
```diff
@@ -273,11 +302,152 @@ func (p *Parser) parseGoFullName() (FullName, bool, error) {
+			return nil, false, err
+		}
+
+		// Case 2: Multiple targets e.g. "&(Person.name, Person.id)".
```
I believe the agreement about the syntax was:
```
(&Person.name, &Person.id)
```

Aflynn50 replied:
I've changed the parser to recognise this. It's working all fine so it seems good to switch to this syntax.

niemeyer reviewed (changes requested):
Thanks, here is a first pass:

Aflynn50 committed:
Add output parser and update tests

Remove unnecessary var decl from parseTargets

Remove code relating to M type

Change []FullName{} to nil in return values when ok=false

Change op, ip and sp to out in and bypass

Remove *.* as a valid column

Remove extraneous skipSpaces

Standardise spacing around checkpoint functions

Make comments consistant and remove named return values from
parseOutput

Change comment wording

Make comments more concise

Address Gustavo comments

Changed sytanx from &(P.a, P.b) to (&P.a, &P.b)

Aflynn50 committed:
Minor refactor and comment update

Aflynn50 committed:
Updated output parser with go style guide advice

niemeyer commented on internal/expr/parser.go:
```diff
@@ -265,28 +272,174 @@ func (p *Parser) parseIdentifier() (string, bool) {
+}
+
+func (p *Parser) parseTarget() (fullName, bool, error) {
+	cp := p.save()
```
Save and restore are not needed in this method given the logic below, and if the input ($) method looks the same, it would be the same.

Aflynn50 replied:
Done.

niemeyer commented on internal/expr/parser.go:
```diff
@@ -265,28 +272,174 @@ func (p *Parser) parseIdentifier() (string, bool) {
+
 // parseGoFullName parses a Go type name qualified by a tag name (or asterisk)
-// of the form "TypeName.col_name". On success it returns the parsed fullName,
+// of the form "&TypeName.col_name". On success it returns the parsed FullName,
```
Curious.. this one was already lowercased, and should probably stay that way.

Aflynn50 replied:
Done.

Aflynn50 committed:
Change comment and remove useless cp.save

niemeyer commented on internal/expr/parser.go:
```diff
@@ -265,28 +272,174 @@ func (p *Parser) parseIdentifier() (string, bool) {
-			return fullName{}, false,
-				fmt.Errorf("invalid identifier near char %d", p.pos)
+		if !p.skipByte('.') {
+			return fullName{}, false, fmt.Errorf("go object near char %d not qualified", p.pos)
```
In a previous review in this PR there was a suggestion to tune these error messages using a `column %d:` sort of prefix, but some of them seem to have left behind. "go" should also be capitalized, but actually we can probably avoid using the term "go" here and instead just talk about types and fields when referring to Go, and tables and columns when talking about the database.

Aflynn50 replied:
Done.

Aflynn50 committed:
Change 'near char x' to 'column x'

niemeyer commented on internal/expr/parser.go:
```diff
@@ -265,28 +272,174 @@ func (p *Parser) parseIdentifier() (string, bool) {
+// parseList takes a parsing function that returns a fullName and parses a
+// bracketed, comma seperated, list.
+func (p *Parser) parseList(parseFn func(p *Parser) (fullName, bool, error)) ([]fullName, bool, error) {
+	cp := p.save()
```
The save can be moved past the first skip below, as a small optimization that prevents doing anything unless this is indeed looking like a proper start at all.

Aflynn50 replied:
I've just realised we can't do this. Because the skipByte is negated, if we **don't** exit the function at the if then the state will have been changed, we will have skipped the `(` byte. So we would be saving the state after the byte has been skipped, thus when we restore it it would not be a true restore.

niemeyer replied:
Good catch. :+1:

Aflynn50 committed:
Optimise cp.save in parseList

niemeyer commented on internal/expr/parser.go:
```diff
@@ -265,28 +272,174 @@ func (p *Parser) parseIdentifier() (string, bool) {
+// parseColumns parses a list of columns. For lists of more than one column the
+// columns must be enclosed in brackets e.g. "(col1, col2) AS &Person.*".
+func (p *Parser) parseColumns() ([]fullName, bool) {
+	cp := p.save()
```
I believe we don't need to save/restore here either, because the entire functionality is implemented inside the two functions below, and only one of them can possibly work.

Aflynn50 replied:
Changed.

niemeyer commented on internal/expr/parser.go:
```diff
@@ -265,28 +272,174 @@ func (p *Parser) parseIdentifier() (string, bool) {
+// If the ampersand is not preceded by a space and succeeded by a name or
+// opening bracket the targets are not parsed.
+func (p *Parser) parseTargets() ([]fullName, bool, error) {
+	cp := p.save()
```
Same here.

Aflynn50 replied:
Changed.

niemeyer commented on internal/expr/parser.go:
```diff
@@ -265,28 +272,174 @@ func (p *Parser) parseIdentifier() (string, bool) {
+// parseOutputExpression parses all output expressions. The ampersand must be
+// preceded by a space and followed by a name byte.
+func (p *Parser) parseOutputExpression() (*outputPart, bool, error) {
+	cp := p.save()
```
Similar here, but here we actually need the save/restore before the parseColumsn branch, since we're moving the pointer forward there and may still fail for other reasons.

Aflynn50 replied:
Changed.

niemeyer reviewed (changes requested):
Good changes, thanks. Second pass, and probably last one.

Aflynn50 committed:
Removed unneeded cp.save/cp.restore

Aflynn50 committed:
Remove extraneous comments

Aflynn50 committed:
Add 'to' in comment

Aflynn50 committed:
Change comments

Aflynn50 committed:
Swaped order of conditional and changed comments

Aflynn50 committed:
Moved cp.save down parseOutputExpression

niemeyer reviewed (approved):
Thanks for the changes!
